### PR TITLE
#1074 in-memory migrations for meeting import

### PR DESCRIPTION
**Pull request by r-peschke**
Dependency: https://github.com/OpenSlides/openslides-datastore-service/pull/188 okay

State: Everything concerning the migration seems okay, including the migration of an OS3-exported database with 150MB JSON-file.
But after migrations there is always the consistence checker, complaining about empty fields with default values and other problems. First one is solved within the checker and a repair-parameter.

### CODE REVIEW

File: openslides_backend/action/actions/meeting/import_.py [hunk 1]
```diff
@@ -55,6 +68,13 @@ def perform(
         self.index = 0
         action_data = self.get_updated_instances(action_data)
         instance = next(iter(action_data))
+
+        if self.name == "meeting.import":
```
Comment (jsangmeister) [inline openslides_backend/action/actions/meeting/import_.py:72]: I'd rather move this into one method which is called every time and can be overriden to do nothing in the subclass

File: openslides_backend/action/actions/meeting/import_.py [hunk 2]
```diff
@@ -131,11 +169,11 @@ def check_usernames_and_generate_new_ones(self, json_data: Dict[str, Any]) -> No
         used_usernames = set()
```
Comment (jsangmeister) [inline openslides_backend/action/actions/meeting/import_.py:169]: Can you unify this method with the one in `user.create` so the behaviour does not diverge again? If you want, we can also do this in a seperate issue to not blow up this one even more

File: openslides_backend/action/actions/meeting/import_.py [hunk 3]
```diff
@@ -430,3 +519,91 @@ def check_permissions(self, instance: Dict[str, Any]) -> None:
...
+        self, instance: Dict[str, Any]
+    ) -> List[CreateEvent]:
+        json_data = instance["meeting"]
+        import_create_events = []
+        for collection in json_data:
+            for entry in json_data[collection].values():
+                # Necessary to remove None-values in event
+                for k, v in list(entry.items()):
+                    if v is None:
+                        entry.pop(k)
```
Comment (jsangmeister) [inline openslides_backend/action/actions/meeting/import_.py:533]: Why? Does the datastore throw an error in that case? Iterating over all entries takes a lot of time so we should only do that if really necessary

File: openslides_backend/action/actions/meeting/import_.py [hunk 4]
```diff
@@ -430,3 +519,91 @@ def check_permissions(self, instance: Dict[str, Any]) -> None:
...
+
+    def create_instance_from_migrated_events(
+        self, instance: Dict[str, Any], migrated_events: List[BaseEvent]
+    ) -> Dict[str, Any]:
+        data: Dict[str, Dict] = defaultdict(dict)
+        for event in migrated_events:
+            collection, id_ = collection_and_id_from_fqid(event.fqid)
+            if event.type == CreateEvent.type:
+                data[collection].update({str(id_): event.data})
+            elif event.fqid.split("/")[0] in ("organization", "committee", "user"):
```
Comment (jsangmeister) [inline openslides_backend/action/actions/meeting/import_.py:602]: Use `collection_from_fqid` here (or at least the `KEYSEPARATOR` constant)

**Review by jsangmeister — CHANGES_REQUESTED** [review]
Looks good! did not check the tests thoroughly. I would like to have this tested with real data to assert that it works as intended. Can you export a real OS3 instance with the latest update where the MI was added and try to import it?

File: openslides_backend/action/actions/meeting/import_.py (hunk 1 above)
Comment (r-peschke) [reply openslides_backend/action/actions/meeting/import_.py:72]: done

File: openslides_backend/action/actions/meeting/import_.py (hunk 3 above)
Comment (r-peschke) [reply openslides_backend/action/actions/meeting/import_.py:533]: The BaseEvent in datastore.migartions calls check and the _ModelEvent implements a check, that a field may not have a None value and raises a BadEventException.

File: openslides_backend/action/actions/meeting/import_.py (hunk 4 above)
Comment (r-peschke) [reply openslides_backend/action/actions/meeting/import_.py:602]: > Use `collection_from_fqid` here (or at least the `KEYSEPARATOR` constant)

done

File: openslides_backend/action/actions/meeting/import_.py (hunk 2 above)
Comment (r-peschke) [reply openslides_backend/action/actions/meeting/import_.py:169]: done

**Pull request merged**